### PR TITLE
feature/Rest-template

### PR DESCRIPTION
**Pull request by ErnestasSeminogovas**
Remove feign dependencies
Remove feign related configurations and services
Rename package feign -> itunes
Add ItunesGateway for endpoint

### CODE REVIEW

File: src/main/java/com/demo/top/itunes/ItunesService.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    public ArtistSearchResponse findArtist(String artistName) {
+        return itunesGateway.getArtists(artistName);
+    }
+
+    public AlbumSearchResponse findTopAlbumsForArtist(Long amgArtistId, Integer limit) {
+        return itunesGateway.getAlbums(amgArtistId, limit);
+    }
+
+    @Autowired
+    public void setItunesGateway(ItunesGateway itunesGateway) {
```
Comment (zilvinasj) [inline src/main/java/com/demo/top/itunes/ItunesService.java]: Why not use constructor injection?

File: src/main/java/com/demo/top/itunes/ItunesGateway.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    try {
+      t = objectMapper.readValue(json, clazz);
+    } catch (Exception ex) {
+      throw new InternalServerException(ex.getMessage());
+    }
+    return t;
+  }
+
+  @Autowired
+  public void setRestTemplate(RestTemplate restTemplate) {
```
Comment (zilvinasj) [inline src/main/java/com/demo/top/itunes/ItunesGateway.java]: Why not use constructor injection?

File: src/main/java/com/demo/top/itunes/ItunesGateway.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  public AlbumSearchResponse getAlbums(Long amgArtistId, Integer limit) {
+    ResponseEntity<String> responseEntity =  restTemplate.getForEntity(itunesUrl + "/lookup?entity=album&amgArtistId=" + amgArtistId + "&limit=" + limit, String.class);
+    return jsonToObject(responseEntity.getBody(),AlbumSearchResponse.class);
+  }
+
+  private <T> T jsonToObject(String json, Class<T> clazz) {
+    if (json == null || json.length() == 0) {
+      throw new InternalServerException("no response body");
```
Comment (zilvinasj) [inline src/main/java/com/demo/top/itunes/ItunesGateway.java]: If the response string is empty it's likely an error on itunes side (therefore not an internal error)

File: src/main/java/com/demo/top/itunes/ItunesService.java (hunk 1 above)
Comment (ErnestasSeminogovas) [reply src/main/java/com/demo/top/itunes/ItunesService.java]: Done

File: src/main/java/com/demo/top/itunes/ItunesGateway.java (hunk 2 above)
Comment (ErnestasSeminogovas) [reply src/main/java/com/demo/top/itunes/ItunesGateway.java]: Done

File: src/main/java/com/demo/top/itunes/ItunesGateway.java (hunk 3 above)
Comment (ErnestasSeminogovas) [reply src/main/java/com/demo/top/itunes/ItunesGateway.java]: Which error code do you suggest then?

Comment (ErnestasSeminogovas) [reply src/main/java/com/demo/top/itunes/ItunesGateway.java]: Changed to FailedDependencyException

File: src/main/java/com/demo/top/itunes/ItunesGateway.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  private ObjectMapper objectMapper;
+
+  @Autowired
+  public ItunesGateway(RestTemplate restTemplate,
+      ObjectMapper objectMapper) {
+    this.restTemplate = restTemplate;
+    this.objectMapper = objectMapper;
+  }
+
+  @Value("${top-application.itunes.url}")
```
Comment (zilvinasj) [inline src/main/java/com/demo/top/itunes/ItunesGateway.java]: this can also be injected through constructor

**Review by zilvinasj — APPROVED** [review]

File: src/main/java/com/demo/top/itunes/ItunesGateway.java (hunk 4 above)
Comment (ErnestasSeminogovas) [reply src/main/java/com/demo/top/itunes/ItunesGateway.java]: Done

**Pull request merged**